File: OrdenamientoMergeSort.py
import time


def _merge(arr, left, mid, right):
    left_half = arr[left: mid + 1]
    right_half = arr[mid + 1: right + 1]

    left_half.append(float('inf'))
    right_half.append(float('inf'))

    i = j = 0
    for k in range(left, right + 1):
        if left_half[i] <= right_half[j]:
            arr[k] = left_half[i]
            i += 1
        else:
            arr[k] = right_half[j]
            j += 1


def _merge_sort_recursive(arr, left, right):
    if left < right:
        mid = (left + right) // 2
        _merge_sort_recursive(arr, left, mid)
        _merge_sort_recursive(arr, mid + 1, right)
        _merge(arr, left, mid, right)


def merge_sort(arr):
    if len(arr) > 1:
        _merge_sort_recursive(arr, 0, len(arr) - 1)


def leer_archivo(nombre_archivo, n):
    try:
        with open(nombre_archivo, 'r') as archivo:
            todos_los_numeros = [int(num) for num in archivo.read().split()]
            return todos_los_numeros[:n]
    except FileNotFoundError:
        print(f"Error al abrir {nombre_archivo}.")
        return None
    except ValueError:
        print("Error: El archivo contiene datos no numéricos.")
        return None


def main():
    nombre_archivo = "datos.txt"

    try:
        n_str = input("Ingrese la cantidad de numeros a leer: ")
        n = int(n_str)
    except ValueError:
        print("Error: Ingrese un número entero válido.")
        return

    if n <= 0:
        print("El valor de n debe ser positivo.")
        return

    arreglo = leer_archivo(nombre_archivo, n)

    if arreglo is None:
        return

    print("\nDatos antes de ordenar:")
    print(*arreglo)

    inicio = time.perf_counter()
    merge_sort(arreglo)
    fin = time.perf_counter()

    tiempo = fin - inicio

    print("\nDatos ordenados:")
    print(*arreglo)

    print(f"\nTiempo de ejecucion Merge Sort: {tiempo:.6f} segundos")


if __name__ == "__main__":
    main()

File: OrdenamientoMergeSort.c
// Merge Sort

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <limits.h>

void merge(int *A, int ini, int mit, int fin) {
    int tizq = mit - ini + 1;
    int tder = fin - mit;
    int *I = (int*) malloc((tizq + 1) * sizeof(int));
    int *D = (int*) malloc((tder + 1) * sizeof(int));

    for (int i = 0; i < tizq; i++) I[i] = A[ini + i];
    for (int j = 0; j < tder; j++) D[j] = A[mit + 1 + j];

    I[tizq] = INT_MAX;
    D[tder] = INT_MAX;

    int i = 0, j = 0;
    for (int k = ini; k <= fin; k++) {
        if (I[i] <= D[j]) {
            A[k] = I[i++];
        } else {
            A[k] = D[j++];
        }
    }

    free(I);
    free(D);
}

void merge_sort(int *A, int ini, int fin) {
    if (ini < fin) {
        int mit = (ini + fin) / 2;
        merge_sort(A, ini, mit);
        merge_sort(A, mit + 1, fin);
        merge(A, ini, mit, fin);
    }
}

int* leerArchivo(const char* nombreArchivo, int tamano, int* tamanoReal) {
    FILE* archivo = fopen(nombreArchivo, "r");
    if (!archivo) {
        printf("Error al abrir %s.\n", nombreArchivo);
        return NULL;
    }

    int* arreglo = (int*) malloc(tamano * sizeof(int));
    if (!arreglo) {
        fclose(archivo);
        printf("Error al asignar memoria.\n");
        return NULL;
    }

    int num, count = 0;
    while (fscanf(archivo, "%d", &num) == 1 && count < tamano) {
        arreglo[count++] = num;
    }

    fclose(archivo);
    *tamanoReal = count;
    return arreglo;
}

int main() {
    int n, tamanoReal;
    char nombreArchivo[] = "datos.txt";

    printf("Ingrese la cantidad de numeros a leer: ");
    scanf("%d", &n);

    if (n <= 0) {
        printf("El valor de n debe ser positivo.\n");
        return 1;
    }

    int* arreglo = leerArchivo(nombreArchivo, n, &tamanoReal);
    if (!arreglo) return 1;

    printf("\nDatos antes de ordenar:\n");
    for (int i = 0; i < tamanoReal; i++) printf("%d ", arreglo[i]);
    printf("\n");

    clock_t inicio = clock();
    merge_sort(arreglo, 0, tamanoReal - 1);
    clock_t fin = clock();
    double tiempo = (double)(fin - inicio) / CLOCKS_PER_SEC;

    printf("\nDatos ordenados:\n");
    for (int i = 0; i < tamanoReal; i++) printf("%d ", arreglo[i]);
    printf("\n");

    printf("\nTiempo de ejecucion Merge Sort: %f segundos\n", tiempo);

    free(arreglo);
    return 0;
}
